### PR TITLE
mqtt dynamic signals

### PR DESCRIPTION
**Pull request by magni-mar**
#428 first
#175 issue

load signals dynamically, tests are included.

### CODE REVIEW

File: exes/mqtt-bridge/inc/run.hpp [hunk 1]
```diff
@@ -27,64 +30,58 @@ class run {
...
-        co_await asio::steady_timer{ sp_interface_.strand(), std::chrono::seconds{ 5 } }.async_wait(asio::use_awaitable);
+      while (!restart_needed_) {
+        co_await asio::steady_timer{ sp_interface_->strand(), std::chrono::milliseconds{ 100 } }.async_wait(
+            asio::use_awaitable);
       }
 
       cancel_signal.emit(asio::cancellation_type::all);
-      tfc_to_exter_.clear_signals();
+
+      io_ctx_.run_for(std::chrono::milliseconds{ 10 });
```
Comment (jbbjarnason) [inline exes/mqtt-bridge/inc/run.hpp:70]: why delay?

**Review by jbbjarnason — COMMENTED** [review]
I am not fond of restart_needed polling, could it be made more event based instead of references and polling?

File: exes/mqtt-bridge/inc/run.hpp (hunk 1 above)
Comment (magni-mar) [reply exes/mqtt-bridge/inc/run.hpp:70]: allow sockets to clear up and connections to close, if this is skipped then the next run will fail

**Comment by magni-mar** [comment]
> I am not fond of restart_needed polling, could it be made more event based instead of references and polling?

I am open to suggestions

File: exes/mqtt-bridge/inc/run.hpp (hunk 1 above)
Comment (jbbjarnason) [reply exes/mqtt-bridge/inc/run.hpp:70]: I see, can you add that to a comment above?

File: exes/mqtt-bridge/inc/run.hpp [hunk 2]
```diff
@@ -27,64 +30,58 @@ class run {
...
-          bind_cancellation_slot(cancel_signal.slot(), asio::detached));
-
-      io_ctx_.run_for(std::chrono::seconds{ 1 });
-
-      while (!restart_needed) {
-        co_await asio::steady_timer{ sp_interface_.strand(), std::chrono::seconds{ 5 } }.async_wait(asio::use_awaitable);
+      while (!restart_needed_) {
+        co_await asio::steady_timer{ sp_interface_->strand(), std::chrono::milliseconds{ 100 } }.async_wait(
+            asio::use_awaitable);
       }
```
[the remark this answers is not in the document]
Comment (magni-mar) [reply exes/mqtt-bridge/inc/run.hpp:66]: Looking into this and will ping this thread when I have a solution, are you talking about `boost::asio::completion_token`?

File: exes/mqtt-bridge/inc/run.hpp (hunk 1 above)
Comment (magni-mar) [reply exes/mqtt-bridge/inc/run.hpp:70]: yes no problem

File: exes/mqtt-bridge/inc/run.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (magni-mar) [reply exes/mqtt-bridge/inc/run.hpp]: This can be implemented for reloading the application based on new signals, there is another reloading which happens when an incorrect packet is received. In my experience, when that happens the program goes into an endless loop and needs to be restarted anyways so I removed that. If that becomes a problem in the future it needs to be investigated on its own.

File: exes/mqtt-bridge/tests/src/integration_tests.cpp [hunk 3]
```diff
@@ -219,46 +220,222 @@ auto main(int argc, char* argv[]) -> int {
...
+    //      timestamp: 1706985768834
+    //      datatype: 11
+    //      is_historical: false
+    //      is_transient: false
+    //      is_null: true
+    //      metadata {
+    //          description: ""
+    //        }
+    //  }
+    //  seq: 0
```
Comment (omarhogni) [inline exes/mqtt-bridge/tests/src/integration_tests.cpp:345]: ?

File: libs/motor/inc/public/tfc/motor/atv320motor.hpp
Comment (omarhogni) [inline libs/motor/inc/public/tfc/motor/atv320motor.hpp:1]: rebase please

File: exes/mqtt-bridge/inc/config/bridge.hpp
Comment (omarhogni) [inline exes/mqtt-bridge/inc/config/bridge.hpp:1]: would have been nice to do this in a separate PR

File: exes/mqtt-bridge/inc/tfc_to_external.hpp [hunk 4]
```diff
@@ -78,19 +93,18 @@ class tfc_to_external {
 
   auto is_publish_signal(std::string signal_name) -> bool {
     // check if signal is a writeable signal
+    for (const auto& banned_signal : config_.value().banned_signals) {
+      if (banned_signal.value == signal_name) {
+        logger_.trace("Signal {} is in the list of banned signals", signal_name);
```
Comment (omarhogni) [inline exes/mqtt-bridge/inc/tfc_to_external.hpp:98]: This can be higher than trace. info would be proper.

**Review by omarhogni — CHANGES_REQUESTED** [review]
there are no tests for new functionality banned_signals.

File: exes/mqtt-bridge/inc/config/bridge.hpp
Comment (magni-mar) [reply exes/mqtt-bridge/inc/config/bridge.hpp:1]: this relates to dynamic loading of signals so that they are discovered when the program is reloaded with new signals, otherwise the config blocks them

**Comment by magni-mar** [comment]
> there are no tests for new functionality banned_signals.

the last test in `integration_tests` is a test for banned_signals

**Review by jbbjarnason — COMMENTED** [review]
This change makes it more easily configurable from my understanding, however, there is a small change which has a rather big impact.

The change of config, from being a whitelist(publish_signals) to a blacklist(banned_signals) this will make it harder to update at customer site. Why this change of course?

What I thought to be the future of this component/interface would be like an option in the IO connections view to enable a signal being published with a simple checkmark or something. I think a whitelist is better for that purpose.

@omarhogni I think we need to make a documentation in regards to future of this components interface and functionality based on our first trial.
Maybe we could use Github Projects for it, if it is beneficial.

**Comment by magni-mar** [comment]
> This change makes it more easily configurable from my understanding, however, there is a small change which has a rather big impact.
> 
> The change of config, from being a whitelist(publish_signals) to a blacklist(banned_signals) this will make it harder to update at customer site. Why this change of course?
> 
> What I thought to be the future of this component/interface would be like an option in the IO connections view to enable a signal being published with a simple checkmark or something. I think a whitelist is better for that purpose.
> 
> @omarhogni I think we need to make a documentation in regards to future of this components interface and functionality based on our first trial. Maybe we could use Github Projects for it, if it is beneficial.

Regardless of weather this function is implemented or not, there are invaluable tests in this PR that should be finished and merged. I can change this PR to tests only and implement dynamic signals. Thoughts?

**Comment by jbbjarnason** [comment]
Let's reopen if we need this feature
